doc/cluster_design.md

```diff
@@ -0,0 +1,137 @@
+# Design: Clustering in SQLflow to analyze patterns in data
+
+## Concept
+
+For analysts and real business people, in the daily analysis work, most of the work is not prediction, but analysis of the patterns in the data. This can help them mine user behavioral characteristics and differences, helping the business discover value and operate.
+
+This design doc introduces how to support the `Cluster Model` in SQLFlow. 
+
+## User interface
+
+In this scenario, we focus on the extraction of data patterns in unsupervised learning. SO, we plan to use a **TRAIN SQL** to train a unsupervised model. We will support whether to perform Pretrain at the beginning of this unsupervised network in `WITH`, and whether to use the already trained model as a pre-training in `USING`. The simple pipeline like:
+
+
+TRAIN SQL:
+
+``` sql
+SELECT * FROM input_table
+TRAIN clusterModel
+WITH
+    model.encode_units = [100, 7]
+    model.n_clusters = 5
+    model.run_pretrain = false
+COLUMN m1, m2, m3, m4, m5, m6, m7, m8, m9, m10 
+USING model.existed_pretrain_model =  existed_pretrain_model
+INTO my_cluster_model;
+```
+
+PREDICT SQL:
+
+``` sql
+SELECT *
+FROM input_table
+PREDICT output_table
+USING my_cluster_model;
+```
+
+where:
+- `input_table` is the high-dimensional table to be clustered.
+- `model.encode_units` is the autoencoder model layer's encoder units, the decode_units can reverse encode_units directly.
+- `model.n_clusters` is the number of patterns after clustering.
+- `my_cluster_model` is the trained cluster model.
+- `run_pretrain`  is used to determine if autoencoder pretrain needs to be run, default true.
+- `model.existed_pretrain_model` is used to specify an existing pretrain_model
+- `output_table` is the cluster result for input_table, which is adding the `group_id` column predicted by the cluster model to the input_table.
+
+## clusterModel Details
+<img src="figures/cluster_model_train_overview.png">
```

Comment:
I think the `decoder` should be include into the stage of `Pre-train`. Because the `auto-encoder` is used for building `encoder` for next training process. The `decoder` will be created at the same time. Even the `decoder` will be never be used in the future, it still should be treated as `Pre-train`(Just my opinion).